From 516ae4e3b8f654103b54bc7e1dc04bf8eff097aa Mon Sep 17 00:00:00 2001
From: Andrei Botila <andrei.botila@nxp.com>
Date: Tue, 18 Mar 2025 09:39:57 +0200
Subject: [PATCH] spi: sja1110: add resume operation for reloading the firmware

After a resume the firmware in the case of SDL boot is no longer
on SJA11110 and has to be written again. Resume is called for both
switch configuration and microcontroller firmware.

Issue: ALBSP-40

Signed-off-by: Andrei Botila <andrei.botila@nxp.com>
---
 sja1110_init.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/sja1110_init.c b/sja1110_init.c
index 4f93fc7..d907869 100644
--- a/sja1110_init.c
+++ b/sja1110_init.c
@@ -33,6 +33,7 @@
 #include <linux/module.h>
 #include <linux/device.h>
 #include <linux/firmware.h>
+#include <linux/pm_runtime.h>
 #include <linux/spi/spi.h>
 #include <linux/of.h>
 #include <linux/of_device.h>
@@ -1115,6 +1116,29 @@ sja1110_remove(struct spi_device *spi)
 #endif
 }
 
+static int __maybe_unused sja1110_suspend(struct device *dev)
+{
+	return 0;
+}
+
+static int __maybe_unused sja1110_resume(struct device *dev)
+{
+	struct spi_device *spi = container_of(dev, struct spi_device, dev);
+	struct sja1110_priv *sja1110 = spi_get_drvdata(spi);
+
+	if (sja1110_init_hw(sja1110))
+		dev_err(&sja1110->spi->dev,
+			"Could not initialize hw (type=%d)\n",
+			sja1110->devtype);
+
+	return 0;
+}
+
+static const struct dev_pm_ops sja1110_pm_ops = {
+	.suspend = sja1110_suspend,
+	.resume = sja1110_resume,
+};
+
 static const struct of_device_id sja1110_dt_ids[] = {
 	{ .compatible = "nxp,sja1110-switch", .data = (const void*)SJA1110_SWITCH },
 	{ .compatible = "nxp,sja1110-uc", .data = (const void*)SJA1110_UC },
@@ -1128,6 +1152,7 @@ static struct spi_driver sja1110_driver = {
 	.driver = {
 		.name = "sja1110",
 		.owner = THIS_MODULE,
+		.pm = &sja1110_pm_ops,
 		.of_match_table = of_match_ptr(sja1110_dt_ids),
 	},
 };
-- 
2.49.0

